Add support for array jobs

benarmston opened the pull request:
This PR adds the initial support for array jobs.

## Array job model

The `Job` model has gained a `job_type` attribute which has three possible values: `JOB`, `ARRAY_JOB` and `ARRAY_TASK`.

A `JOB` represents a non-array job, its submission script is submitted to a single node allocated to it.  This is intended to be the type for interactive jobs too.

An `ARRAY_JOB` represents exactly that, an array job.  When created, it will have a number of `ARRAY_TASK`s created for it.  The array task is the job which is submitted to the daemon to be ran.  The state of the array task can be tracked separately from the array job.

The scheduler does not concern itself with `ARRAY_TASK`, instead it allocates resources and schedules `JOB`s and `ARRAY_JOB`s.  When an `ARRAY_JOB` is ran its `ARRAY_TASK`s will be submitted to the allocated nodes.

## Refactoring of submission and cancellation code

The code dealing with submission and cancellation of jobs has been refactored.  There is now a class `Submission::BatchJob` which encapsulates submitting a job to a single node.  There is `Submission::ArrayTask` which encapsulates submitting a single array task to a single node.  `Submission::EnvGenerator` has been added which encapsulates generating the environment for batch and array jobs.

Job cancellation has seen some similar refactoring; `Cancellation::BatchJob` and `Cancellation::ArrayJob`.

There is still some duplication in `EventProcessor` and the new classes indicating that our models are not quite correct, but they will be dealt with in due course as the correct abstractions reveal themselves.

## Limitations

* Currently, only a single node allocated to an array job is used to run the array tasks, and the tasks are not ran in parallel.  To fix these, we need some way of tracking how many tasks can be ran on per-node and which node has been allocated to which task.  Our allocation and job models do not currently support.

* `ARRAY_JOB`s are not displayed correctly in the output of `queue`.  They need to be distinguished from non-array jobs in some manner.

* Cancelling array jobs is supported, but cancelling individual array tasks is not.

benarmston committed:
Extract Submission::BatchJob and EnvGenerator

The environment generation and batch job submission code have been
extracted to `Submission::BatchJob` and `Submission::EnvGenerator`.
This cleans up some code in `EventProcessor` and will support branching
in the appropriate locations for submission of array jobs/tasks.

benarmston committed:
Initial pass at modelling array job/tasks

There are now three types of jobs: `JOB`, `ARRAY_JOB` and `ARRAY_TASK`.

A `JOB` represents a non-array job, its submission script is submitted
to a single node allocated to it.

An `ARRAY_JOB` represents exactly that, an array job.  When created, it
will have a number of `ARRAY_TASK`s created for it.  The array task is
the job which is submitted to the daemon to be ran.  The state of the
array task can be tracked separately from the array job.

The scheduler does not concern itself with `ARRAY_TASK`, instead it
allocates resources and schedules `JOB`s and `ARRAY_JOB`s.  When an
`ARRAY_JOB` is ran its `ARRAY_TASK`s will be submitted to the allocated
nodes.

Limitations of array tasks:

Currently, only a single node allocated to an array job is used to run
the array tasks, and the tasks are not ran in parallel.  To fix these,
we need some way of tracking how many tasks can be ran on per-node and
which node has been allocated to which task.  Our allocation and job
models do not currently support.

benarmston committed:
Run array jobs/tasks

benarmston committed:
Refactor cancellation of non-array jobs

The job cancellation code has been refactored similarly to the job
submission code.  A new state of `CANCELLING` has been added to jobs.

benarmston committed:
Cancel array jobs/tasks

Array jobs/tasks can be cancelled too.  When cancelling either, the
entire array job is cancelled.

WilliamMcCumstie committed:
Update the WebSocket documentation

WilliamMcCumstie committed:
Document the environment variables

WilliamMcCumstie commented on app.rb:
```diff
@@ -214,14 +214,17 @@ def validate!
       @script = attr[:script]
       job = Job.new(
+        arguments: attr[:arguments],
+        array: attr[:array],
```
Currently the `array` input is missing its documentation which has hidden an idiosyncrasy. The `array` input is a string in csv format where the `arguments` is a literal array of strings.

We need to document this difference to make it easier on the client to craft the request.

It might be worth considering submitting the "arrays" in a standardised format (e.g. either csv or literal json array).

benarmston replied:
The format that slurm supports for the `--array` parameter is not as simple as I've implemented here.  In order to support migration from slurm, we'd like to be backwards compatible on as many options as possible, particularly the more common options.

As such, I expect that `--array` will eventually support much of the slurm syntax.  Some examples are:

* `--array 0-3` == `--array 0,1,2,3`
* `--array 0,1,4-5` == `--array 0,1,4,5`
* `--array 0-15:4` == `--array 0,4,8,12`

This shorthand is obviously not suitable for the `arguments` parameter.

Whilst it might seem tempting to expand the shorthand client-side and then submit the arrays in a common format, that is not suitable for large arrays.  I would much sooner send the string `0-1023` than I would its expanded version.  Likewise, I'd much soon read the string `0-8,10-15,18-23` in the logs than its expanded version.

WilliamMcCumstie replied:
Okay thanks for the clarification. Possible the API attribute `array` not being an array is the source of the confusion. The CLI `--array` input doesn't necessarily need parity as the client can handle the key translation.

Possible `array_range` (or something to that effect) would be clearer. However this is me being :panda_face:

WilliamMcCumstie commented on app/models/job.rb:
```diff
@@ -91,7 +132,21 @@ def script_path
+
+  def create_array_tasks
+    self.array_tasks ||= @array.split(',').map do |idx|
+      self.dup.tap do |task|
```
This looks kinda strange to duplicate the job to generate the tasks. I believe we are conflating `task` and `job` into a single model when they are quite different objects.

Possible it would be better to create multiple `Task` decorators around a single `Job`. This way it would be easier for them to develop in different directions. `SimpleDelegator` would be a fairly easy first pass.

Bear in mind this is only a shallow duplication, so there maybe shared resource issue in the future

benarmston replied:
You might be right about conflating a task object with a job object.  I expect that will become clearer as the remainder of array job/task functionality is implemented.

I've changed `create_array_tasks` to create new `Job`s instead of duplicating and adjusted the validations too.  I don't think you'll like the change any more than you do the existing mechanism, but it is much clearer in what attributes belong to which type of job.

WilliamMcCumstie replied:
Looks like a step in the right direction. My gut instinct is eventually batch submissions would a `job`-`task` in a one-to-one relationship where array submissions are one-to-many.

However this is more LTE of the data model and can be shelved for now.

WilliamMcCumstie commented on app/websocket_app.rb:
```diff
@@ -49,6 +49,16 @@ def call(message)
 
+    when 'NODE_COMPLETED_ARRAY_TASK'
+      task_id = message[:array_task_id]
+      job_id = message[:array_job_id]
```
In general we should be able to infer the `job_id` from the `task_id` as they exist in a one-many relationship. Looks like we need both due to the data model.

This is fine for now as we can deprecate the `job_id` without introducing a breaking change.

benarmston replied:
At the moment, we can't sensibly retrieve an array task from its id.  Once we have job/task persistence in place we'll be able to do so.  At that point, sending only the `task_id` might be the better option.

Regarding, deprecation and breaking changes, we don't need to concern ourselves with that whilst the project is still in alpha.  We just need to make sure that our three `master` branches work together correctly.

WilliamMcCumstie replied:
:+1:

WilliamMcCumstie reviewed (changes requested):
Looks like the right approach to implement the array tasks but I have a few hesitations mentioned below.

I have updated some of the more technical documentation surrounding the websockets. These are particularly important as the end user needs to find out which env vars are set. Whilst this still needs work at least we can tell them to do the following:

```
curl http://localhost/v0/docs | jq .definitions.jobAllocatedWS.properties.environment.properties
```

benarmston committed:
Fix specs

benarmston committed:
Add specs for array jobs

benarmston committed:
Minor refactor of array tasks

This makes it clearer what attributes of ARRAY_TASKs are being set and
used.  The number of branches on `job_type` in the validations
especially might indicate that ARRAY_TASKs should be modeled as a
separate `Task` model.

benarmston commented:
@WilliamMcCumstie did you find time to consider 

* whether there are any obvious blockers with the current implementation for running multiple array tasks on a single node in parallel. The `--ntasks*` options that sbatch uses.

* whether there are any obvious blockers with the current implementation for running multiple array tasks on multiple nodes in parallel. The `--nodes` option in conjunction with the `--array` option.

It would be good to hear your thoughts on how we could support that.

WilliamMcCumstie commented:
In regards to `--nodes` and `--ntasks*` I don't see any immediate issues, however I have identified the following known unknowns:

* How well the `daemon` handles the async processing of multiple tasks (untested),
* How well error handling preforms if a single task fails, and
* Cancelling a single task (possible unsupported?).

Really we wont know until we try, so might as well steam ahead.

WilliamMcCumstie commented:
Rebased to resolve conflicts

WilliamMcCumstie committed:
Fix NoMethod error job.script bug

This was likely introduced during a rebase

WilliamMcCumstie commented:
This LGTM, my final comments can be addressed as part of a future iteration.
